client/cache/cache_pattern.go

```diff
@@ -0,0 +1,110 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+)
+
+// ValueReader reads data from data source, such as database
+type ValueReader interface {
+	Query(key string) (interface{}, error)
+}
+
+// ValueWriter writes data to data source, such as database
+type ValueWriter interface {
+	Update(key string, val interface{}) error
+}
+
+// CacheAside is a decorator of Cache, it implements "cache aside" pattern
+type CacheAside struct {
+	Cache
+	reader   ValueReader
+	writer   ValueWriter
+	lifeSpan time.Duration
+}
+
+func NewCacheAside(cache Cache, reader ValueReader, writer ValueWriter, lifeSpan time.Duration) *CacheAside {
```

Comment:
I think both reader, writer and lifeSpan are optional.

For example, users only want `CacheAside` read data from DB if the cache didn't have the specific key. But users update values by ORM directly:
```go

func demo() {
   aside.NewCacheAside(redisCache, ormReader)
   aside.Get(ctx, key1)
   orm.update(key1, xxx)
}

```

Reply:
> I think both reader, writer and lifeSpan are optional.
> 
> For example, users only want `CacheAside` read data from DB if the cache didn't have the specific key. But users update values by ORM directly:
> 
> ```go
> func demo() {
>    aside.NewCacheAside(redisCache, ormReader)
>    aside.Get(ctx, key1)
>    orm.update(key1, xxx)
> }
> ```

I think `CacheAside` encapsulates operation of data source and cache for user. If the user implements the data source operation by himself, I think it loses the meaning of `CacheAside`.